Adds a few endpoints and the outdated banner

maxlang opened the pull request:
- cluster API endpoint for the cluster ID
- Cockroach Labs version check endpoint

NOTE: missing a few tests

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7339)

<!-- Reviewable:end -->

Commit:
Adds a few endpoints and the outdated banner

- cluster API endpoint for the cluster ID
- Cockroach Labs version check endpoint

tamird commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[ui/next/app/containers/banner/bannerContainer.tsx, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTC9pa8GztPjz3uN4:-KKjTC9pa8GztPjz3uN5:-1398239445) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/bannerContainer.tsx#L13)):*
> ```XML
>   render() {
>     return <div id="banner">
>       {/* Note: the order the banners appear here dictates their precedence */}
> ```

can we make this a code comment? seems odd to put it in the "built" page.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjUwbkIm8qnmG3vy5a:-KKjUwbkIm8qnmG3vy5b:197934709) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L15)):*
> ```XML
> import _ = require("lodash");
> 
> const DAY = 1000 * 60 * 60 * 24;
> ```

MILLISECONDS_IN_DAY?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW2VCFf2RJxx8IR6w:-KKjW2VCFf2RJxx8IR6x:-769440920) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L70)):*
> ```XML
>   bannerText() {
>     if (!this.props.versionsMatch) {
>       return "The node versions are mismatched.";
> ```

does this want to list the versions?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW4DO--TRvRmLzSui:-KKjW4DO--TRvRmLzSuj:-769440920) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L72)):*
> ```XML
>       return "The node versions are mismatched.";
>     } else if (this.props.newerVersions) {
>       return "There is a newer version of CockroachDB available.";
> ```

does this want to list the versions?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWY3RWISS5awV2AX2:-KKjWY3RWISS5awV2AX3:-1700213615) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L77)):*
> ```XML
> 
>   visible() {
>     // check if it was locally dismissed
> ```

what does "locally" mean in this case?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWdBC-kRuKY4Z9AGU:-KKjWdBDF7wdwOvQbNog:1513035490) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L82)):*
> ```XML
>       (this.props.versionCheckDismissedFetched && !this.props.versionCheckDismissed || (Date.now() - this.props.versionCheckDismissed > DAY)) &&
>       // check if it would show any text (only if there's a version mismatch or a newer version)
>       !!this.bannerText();
> ```

yikes. how about `this.bannerText() !== null`?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 101 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjX7BjYAPcwYv02sjx:-KKjX7BkR07xzITU8bR4:1632537229) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L101)):*
> ```XML
> let clusterID = (state: any): string => state.cluster.data && state.cluster.data.cluster_id;
> 
> // TODO: change the way uiData stores data so we don't need an added check to see if the value was fetched
> ```

put your name on these TODOs

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 104 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXOSRcxDAU6KPNQ-w:-KKjXOSRcxDAU6KPNQ-x:-1437200386) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L104)):*
> ```XML
> // TODO: determine whether they dismissed a node version mismatch or an upstream version check
> let versionCheckDismissedFetched = (state: any): boolean => state && state.uiData && state.uiData.data && _.has(state.uiData.data, VERSION_DISMISSED_KEY);
> let versionCheckDismissed = (state: any): number => state && state.uiData && state.uiData.data && state.uiData.data[VERSION_DISMISSED_KEY];
> ```

can't this be `versionCheckDismissedFetched(state) && state.uiData.data[VERSION_DISMISSED_KEY];`?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXb1yVq5_z3-7FGJi:-KKjXb1yVq5_z3-7FGJj:-2069916680) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L116)):*
> ```XML
> );
> 
> let buildtag = createSelector(
> ```

seems like the wrong place for this kind of logic - i'd expect iteration over a list of versions to happen in just one place.

also, don't the versions want to be tracked in a set? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTLNJ1o6wPaalFmjw:-KKjTLNJ1o6wPaalFmjx:-242988457) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/interfaces/cockroachlabs.d.ts#L1)):*
> ```XML
> export interface Version {
> ```

this file needs a comment that points to the upstream data structures it's tracking.

---

*[ui/next/app/redux/cluster.ts, line 2 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjY14lx1emnpU63MRW:-KKjY14lx1emnpU63MRX:1842223780) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L2)):*
> ```XML
> /**
>  * This module maintains the state of read-only data about the cluster
> ```

missing period here and on the next line.

---

*[ui/next/app/redux/cluster.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXpkTQfimdotBasrO:-KKjXpkU0ysYTAIyPjpf:1213132838) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L28)):*
> ```XML
>  * Redux reducer which processes actions related to the cluster query.
>  */
> export default function reducer(state: ClusterState = new ClusterState(), action: Action): ClusterState {
> ```

meta comment: this is copy-pasta'd around a lot.

---

*[ui/next/app/redux/cluster.ts, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXtvcX_XwSj0U-jDm:-KKjXtvcX_XwSj0U-jDn:-1114887027) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L96)):*
> ```XML
>     let { cluster }: {cluster: ClusterState} = getState();
> 
>     // TODO: Currently we only refresh when the page is revisited. Eventually
> ```

put your name on this.

---

*[ui/next/app/redux/uiData.ts, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTa5BOQFmZSGo7CBn:-KKjTa5BOQFmZSGo7CBo:470097438) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/uiData.ts#L32)):*
> ```XML
> }
> 
> // version check key
> ```

throughout: can we adopt similar comment styles in the UI as we have on the server? complete sentences on non-inline comments, in this case.

---

*[ui/next/app/redux/version.ts, line 2 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXz4LXBr2D4Lr0BSR:-KKjXz4LXBr2D4Lr0BSS:1672428078) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/version.ts#L2)):*
> ```XML
> /**
>  * This module maintains the state of read-only data about version currency
> ```

missing period.

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjTkJXUU7t23oxBpkR:1456646637) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
> ```XML
> 
> // getHealth gets health data
> export function getHealth(req: HealthRequest = {}): Promise<HealthResponseMessage> {
> ```

I know this is equivalent, but let's leave the old. We may have a payload in the future, and there's no harm in leaving the optional argument.

ditto below.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKjTwuSbaPVKIE6f7QC:-1934956816) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
> ```XML
> export const COCKROACHLABS_ADDR = "https://register.cockroachdb.com";
> 
> function timeoutFetch<T extends BodyInit>(url: string, req?: T) {
> ```

this is almost exactly `api`'s `fetch`. Do you think you could extract a common function?

also plz include return type (use another generic parameter rather than `any`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Reviewed 2 of 9 files at r1.
Review status: all files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[ui/next/app/containers/banner/bannerContainer.tsx, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTC9pa8GztPjz3uN4:-KKj_y-dlckfZzEP6fGc:-51345028) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/bannerContainer.tsx#L13)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can we make this a code comment? seems odd to put it in the "built" page.
</details>
This is how you put comments into JSX (breaking out into a code block), but in this case I agree it can be moved up. 

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjUwbkIm8qnmG3vy5a:-KKj_mvIIr5UhiRn5FX2:-1870528502) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> MILLISECONDS_IN_DAY?
</details>
Also consider using something like moment.duration(1, 'd')

---

*[ui/next/app/redux/cluster.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXpkTQfimdotBasrO:-KKjbqVnLenatH0W1lup:1063155241) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> meta comment: this is copy-pasta'd around a lot.
</details>
There is a massive opportunity for a "local data cache" component; that's what more than half of our reducers represent, read-only periodically refreshed data.

It's a bit tricky to make reusable reducers, but it can be done.

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjbMQtMiTMRUjOqZL6:221612879) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I know this is equivalent, but let's leave the old. We may have a payload in the future, and there's no harm in leaving the optional argument.
> 
> ditto below.
</details>
But `req` is not even being used.  Are you suggesting to keep this here as some sort of suggestion on the future planned direction of the API?  I don't know of any specific plans for a payload in this function.  I say remove it.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKjabMC6ZHqVfeq-KZo:467644394) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is almost exactly `api`'s `fetch`. Do you think you could extract a common function?
> 
> also plz include return type (use another generic parameter rather than `any`)
</details>
The return value of res.json() is always going to be `any`, we don't have control of that library.  In this case, it's already being cast by `versionCheck`, which specifies a return type of `Promise<VersionList>`.  I don't understand the value of casting it any earlier than that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWdBC-kRuKY4Z9AGU:-KKjdKyGNQOhJHqif6sh:-35791310) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> yikes. how about `this.bannerText() !== null`?
</details>
That would be true for `undefined`.  `!!` is the most concise way to determine if a javascript value is "truthy".

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjeCVUHoffs7uCNeN2:-KKjeCVUHoffs7uCNeN3:-977415137) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L103)):*
> ```XML
> // TODO: change the way uiData stores data so we don't need an added check to see if the value was fetched
> // TODO: determine whether they dismissed a node version mismatch or an upstream version check
> let versionCheckDismissedFetched = (state: any): boolean => state && state.uiData && state.uiData.data && _.has(state.uiData.data, VERSION_DISMISSED_KEY);
> ```

We should probably just add methods to the UIDataState object, this pattern is being used in multiple places.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjcEJluvs1hAfQttXh:-KKjcEJmY2BpJUOp34FX:-1321326865) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/interfaces/cockroachlabs.d.ts#L16)):*
> ```XML
> }
> 
> interface RegisterationData {
> ```

s/RegisterationData/RegistrationData/g

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[ui/next/app/containers/banner/bannerContainer.tsx, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTC9pa8GztPjz3uN4:-KKjYmG9Ijq7AGuTbXhA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/bannerContainer.tsx#L13)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can we make this a code comment? seems odd to put it in the "built" page.
</details>
Done.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjUwbkIm8qnmG3vy5a:-KKjYnDa8LePU9YjeRTW:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> MILLISECONDS_IN_DAY?
</details>
Done.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW2VCFf2RJxx8IR6w:-KKjZEk4gntXzUmCVxVg:-1498449308) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this want to list the versions?
</details>
Changed to be consistent with the old code.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW4DO--TRvRmLzSui:-KKjYtxSLBPez_K_0sjR:1294824757) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does this want to list the versions?
</details>
This is consistent with the old code. Early on I had text that read "There are X newer versions of Cockroach Labs available" but it was suggested that this was simpler at some point. I have no strong preference, but I think we should probably annotate updates at some

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWY3RWISS5awV2AX2:-KKjZNuv1Bt-RyzVFO8_:-1158617022) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what does "locally" mean in this case?
</details>
When you click the close button we both store that dismissal both in the client, and on the server. This allows the banner to be hidden immediately, even if the network has issues.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWdBC-kRuKY4Z9AGU:-KKjlZuxLVrZ3QqaVz0H:1828934179) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L82)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> That would be true for `undefined`.  `!!` is the most concise way to determine if a javascript value is "truthy".
</details>
`this.bannerText() != null` or `this.bannerText() !== undefined` would work, but I agree with @mrtracy that this is more concise. It's essentially casting to a boolean.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 101 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjX7BjYAPcwYv02sjx:-KKjaz2XuzkdyjGimu3-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> put your name on these TODOs
</details>
Done.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 103 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjeCVUHoffs7uCNeN2:-KKjl5pSd5May4d9DygM:907812329) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L103)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> We should probably just add methods to the UIDataState object, this pattern is being used in multiple places.
</details>
Yeah, that's why I added the todo.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 104 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXOSRcxDAU6KPNQ-w:-KKjb17tADCHTbNUzQVG:-1478840880) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can't this be `versionCheckDismissedFetched(state) && state.uiData.data[VERSION_DISMISSED_KEY];`?
</details>
Actually, I've since realized that state and state.uiData will never be missing, so I'll simplify these.

You'd have to write it `versionCheckDismissedFetched(state) ? state.uiData.data[VERSION_DISMISSED_KEY] : null` to ensure the type is correct.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXb1yVq5_z3-7FGJi:-KKjc0E2As2ucodmeYwX:-697926447) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L116)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems like the wrong place for this kind of logic - i'd expect iteration over a list of versions to happen in just one place.
> 
> also, don't the versions want to be tracked in a set? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
</details>
The iteration does happen just once, in the `versions` selector. That said, I don't use the counts of nodes per version anymore, so I'll just use `_.map` and `_.uniq` instead.

I've been shying away from `Set` because it doesn't seem well supported by IE or Safari.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTLNJ1o6wPaalFmjw:-KKjehQ6i0DI84de28C8:-1556522279) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/interfaces/cockroachlabs.d.ts#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this file needs a comment that points to the upstream data structures it's tracking.
</details>
I'm not sure how we want to do this, since it points to code in a private repo.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 16 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjcEJluvs1hAfQttXh:-KKjlUCBCbbdrLdkomYL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/interfaces/cockroachlabs.d.ts#L16)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> s/RegisterationData/RegistrationData/g
</details>
Done.

---

*[ui/next/app/redux/cluster.ts, line 2 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjY14lx1emnpU63MRW:-KKjlYxClzN0OFXTFG5N:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L2)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> missing period here and on the next line.
</details>
Done.

---

*[ui/next/app/redux/cluster.ts, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXpkTQfimdotBasrO:-KKjeogr_NIhdr34IAF3:-815817403) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L28)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> There is a massive opportunity for a "local data cache" component; that's what more than half of our reducers represent, read-only periodically refreshed data.
> 
> It's a bit tricky to make reusable reducers, but it can be done.
</details>
Yeah, I've been wanting to do it, but was wary of the time it would take, given that we want to switch over to next. As soon as we make the switch I'm going to refactor a bit to avoid all the copy-paste.

---

*[ui/next/app/redux/cluster.ts, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXtvcX_XwSj0U-jDm:-KKjf60mYc1-53BeWPFO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/cluster.ts#L96)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> put your name on this.
</details>
Done.

---

*[ui/next/app/redux/uiData.ts, line 32 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTa5BOQFmZSGo7CBn:-KKjfT2p5OWBCdxRDMfd:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/uiData.ts#L32)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: can we adopt similar comment styles in the UI as we have on the server? complete sentences on non-inline comments, in this case.
</details>
Done

---

*[ui/next/app/redux/version.ts, line 2 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXz4LXBr2D4Lr0BSR:-KKjf_1xWQ7yoVre8DTj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/redux/version.ts#L2)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> missing period.
</details>
Done.

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjfcAj_K0lQ6hBP78m:-367319385) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> But `req` is not even being used.  Are you suggesting to keep this here as some sort of suggestion on the future planned direction of the API?  I don't know of any specific plans for a payload in this function.  I say remove it.
</details>
I decided to take it out because it's confusing if you look around to see where `req` is being used, even though I liked the symmetry. We'll have the change the health request anyway (to a POST) if it ever does need a payload, so I think it will just mislead someone looking at the codebase.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKjguogj-4CAmcQzSVK:-1174179870) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> The return value of res.json() is always going to be `any`, we don't have control of that library.  In this case, it's already being cast by `versionCheck`, which specifies a return type of `Promise<VersionList>`.  I don't understand the value of casting it any earlier than that.
</details>
The biggest issue is the generics. I'm not super familiar with doing function overloads in typescript, which is what you'd need here (basically if the builder is missing, you assume it's the current `timeoutFetch`. I'll add a TODO to revisit it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 2 of 10 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWdBC-kRuKY4Z9AGU:-KKjkqJTiAvQWj7TC6Wa:1003196960) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L82)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> That would be true for `undefined`.  `!!` is the most concise way to determine if a javascript value is "truthy".
</details>
err, sorry, you should compare to undefined, since apparently a missing return results in undefined.

I don't think checking for "truthy" is appropriate here; we have type information, we don't need to use type-clobbering constructs like `!!`

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjlDTqiU3LDchX7LqG:-1602490273) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> But `req` is not even being used.  Are you suggesting to keep this here as some sort of suggestion on the future planned direction of the API?  I don't know of any specific plans for a payload in this function.  I say remove it.
</details>
Yes, that's what I'm suggesting. This is defined in terms of a proto - why would you burden yourself with the decision of when to include this parameter vs when to exclude it? It's just as simple to include it always.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKjlQLHDqlIMaYNlYyq:-1771060580) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> The biggest issue is the generics. I'm not super familiar with doing function overloads in typescript, which is what you'd need here (basically if the builder is missing, you assume it's the current `timeoutFetch`. I'll add a TODO to revisit it.
</details>
`.json()` returns `Promise<any>`, but `.json<T>()` returns `Promise<T>`.  By plumbing the type `T` through, you can prevent future mistakes where this function (`timeoutFetch()`, not `json()`) returns the wrong thing. For instance (strawman), it would prevent the trivial error of `return Promise.resolve({})`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[ui/next/app/containers/banner/bannerContainer.tsx, line 12 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjmBdz04Z8YPBpWZ4A:-KKjmBe-jfD9SGNY-CAI:481852382) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17cf6849aebf69cc4ae6cdb0a7c2362f7771953/ui/next/app/containers/banner/bannerContainer.tsx#L12)):*
> ```XML
> export default class extends React.Component<{}, {}> {
>   render() {
>     // Note: the order the banners appear here dictates their precedence
> ```

please add a period.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW2VCFf2RJxx8IR6w:-KKjmGfEuXtYOVdeVrg7:931963071) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L70)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Changed to be consistent with the old code.
</details>
I didn't say it should be consistent with the old code - I asked if it should list the versions. Should it? It seems like we're now making the operator scratch their head to figure out which versions are running on which nodes.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW4DO--TRvRmLzSui:-KKjmRC6IsfKSlkqdNvU:-283747606) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L72)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> This is consistent with the old code. Early on I had text that read "There are X newer versions of Cockroach Labs available" but it was suggested that this was simpler at some point. I have no strong preference, but I think we should probably annotate updates at some
</details>
Again, this is making more work for the operator. Which version am I on? What's the latest version? Where are the release notes? etc.

We can do better about providing some breadcrumbs here without being overly verbose.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWY3RWISS5awV2AX2:-KKjmf-6a4MYPgoPgS7L:-1576674434) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L77)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> When you click the close button we both store that dismissal both in the client, and on the server. This allows the banner to be hidden immediately, even if the network has issues.
</details>
OK, can you clarify the comment please? Also, these comments should look like sentences.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWdBC-kRuKY4Z9AGU:-KKjmqVZPuyE8yW5Nm20:-1712046658) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> err, sorry, you should compare to undefined, since apparently a missing return results in undefined.
> 
> I don't think checking for "truthy" is appropriate here; we have type information, we don't need to use type-clobbering constructs like `!!`
</details>
Why do you need to check both null and undefined?

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 104 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXOSRcxDAU6KPNQ-w:-KKjmyj0vwv4n3rofvha:1875355645) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L104)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Actually, I've since realized that state and state.uiData will never be missing, so I'll simplify these.
> 
> You'd have to write it `versionCheckDismissedFetched(state) ? state.uiData.data[VERSION_DISMISSED_KEY] : null` to ensure the type is correct.
</details>
Still seems better to me.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXb1yVq5_z3-7FGJi:-KKjnk-EB2jM4LOXj_X_:-1037760788) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L116)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> The iteration does happen just once, in the `versions` selector. That said, I don't use the counts of nodes per version anymore, so I'll just use `_.map` and `_.uniq` instead.
> 
> I've been shying away from `Set` because it doesn't seem well supported by IE or Safari.
> 
</details>
Are we really targeting IE? Safari 9 seems to support it reasonably well.

As I said elsewhere, I think we should be displaying some additional information about nodes and their versions, which would require grouping by version here.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTLNJ1o6wPaalFmjw:-KKjo2vsbaa6bzCLK3KT:2098241015) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/interfaces/cockroachlabs.d.ts#L1)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I'm not sure how we want to do this, since it points to code in a private repo.
</details>
I'd be OK with a link (it's no secret that there's a private repo).

---

*[ui/next/app/util/api.spec.ts, line 472 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjofRiaQKfa2VcyL5H:-KKjofRiaQKfa2VcyL5I:-1973758406) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17cf6849aebf69cc4ae6cdb0a7c2362f7771953/ui/next/app/util/api.spec.ts#L472)):*
> ```XML
>     it("correctly requests cluster info", function () {
>       this.timeout(1000);
>       // Mock out the fetch query
> ```

useless comment

---

*[ui/next/app/util/api.spec.ts, line 490 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjogY0nn1ddRF4xGXS:-KKjogY0nn1ddRF4xGXT:-880498262) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17cf6849aebf69cc4ae6cdb0a7c2362f7771953/ui/next/app/util/api.spec.ts#L490)):*
> ```XML
>       this.timeout(1000);
> 
>       // Mock out the fetch query, but return a 500 status code
> ```

this comment lies and is useless

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjoM7aAb42KGFUphF9:1226347175) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
> We'll have the change the health request anyway (to a POST) if it ever does need a payload, so I think it will just mislead someone looking at the codebase.

Not true; there are several request types in this file that we map to REST-y GETs. I don't see the value in introducing special cases to the pattern in this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 5 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[ui/next/app/containers/banner/bannerContainer.tsx, line 12 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjmBdz04Z8YPBpWZ4A:-KKjpQgenX2vJ9EJ8JoU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17cf6849aebf69cc4ae6cdb0a7c2362f7771953/ui/next/app/containers/banner/bannerContainer.tsx#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please add a period.
</details>
Done.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjUwbkIm8qnmG3vy5a:-KKjmLlcmaC9ssOTTvsU:481603083) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L15)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Done.
</details>
Just saw this. I like the moment version.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW2VCFf2RJxx8IR6w:-KKjpVFY7XZxNWS1EPRK:-1289801791) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I didn't say it should be consistent with the old code - I asked if it should list the versions. Should it? It seems like we're now making the operator scratch their head to figure out which versions are running on which nodes.
</details>
Maybe, but let's track this in a separate issue. Right now the purpose is to port over the outdatedBanner. I don't think listing out a bunch of version tags or nodes will scale well, especially if the number is large.

https://github.com/cockroachdb/cockroach/issues/7343

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW4DO--TRvRmLzSui:-KKjp_R6G9x0kkr5W_u9:160841994) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L72)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Again, this is making more work for the operator. Which version am I on? What's the latest version? Where are the release notes? etc.
> 
> We can do better about providing some breadcrumbs here without being overly verbose.
</details>
Same as above.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 77 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWY3RWISS5awV2AX2:-KKjtDE_PVqiY_np7vkD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L77)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK, can you clarify the comment please? Also, these comments should look like sentences.
</details>
Done.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 82 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjWdBC-kRuKY4Z9AGU:-KKjs5ikvmBMeX_kY0VY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Why do you need to check both null and undefined?
</details>
Done.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 104 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXOSRcxDAU6KPNQ-w:-KKjsAJz91vaXR-llzsV:2017567679) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Still seems better to me.
</details>
Really? Seems more verbose. I'll change it, but as @mrtracy pointed out this pattern needs fixing anyway.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 116 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjXb1yVq5_z3-7FGJi:-KKjsaSEmrKkwWJTSjOe:-184406554) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L116)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Are we really targeting IE? Safari 9 seems to support it reasonably well.
> 
> As I said elsewhere, I think we should be displaying some additional information about nodes and their versions, which would require grouping by version here.
</details>
Well neither of them seem to support the `new Set(iterator)` syntax which seems the most useful.

Let's address that separately as I mentioned.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTLNJ1o6wPaalFmjw:-KKjt9XBpaottuEAGVrJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/interfaces/cockroachlabs.d.ts#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'd be OK with a link (it's no secret that there's a private repo).
</details>
Done.

---

*[ui/next/app/util/api.spec.ts, line 472 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjofRiaQKfa2VcyL5H:-KKjtBB6341L3bzl3vS7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17cf6849aebf69cc4ae6cdb0a7c2362f7771953/ui/next/app/util/api.spec.ts#L472)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> useless comment
</details>
Done.

---

*[ui/next/app/util/api.spec.ts, line 490 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjogY0nn1ddRF4xGXS:-KKjtBjAlZOwP5yoTHkD:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e17cf6849aebf69cc4ae6cdb0a7c2362f7771953/ui/next/app/util/api.spec.ts#L490)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this comment lies and is useless
</details>
Done.

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjnvr1XmcMbWVZGmUO:-1034827387) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yes, that's what I'm suggesting. This is defined in terms of a proto - why would you burden yourself with the decision of when to include this parameter vs when to exclude it? It's just as simple to include it always.
</details>
If, in the future, the backend changed, you'd still need to add it to the next line as an argument. The fact that it sits here is misleading. It tricks you into thinking you can/should pass something in when you shouldn't and it tricks you into thinking it will be used when it isn't.

If this were an empty POST I would agree, but here it's a get, and any changes on the backend to the Health proto would still require changes to `api.ts`.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKjomBq8A82nxzs8aV0:-1693089994) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `.json()` returns `Promise<any>`, but `.json<T>()` returns `Promise<T>`.  By plumbing the type `T` through, you can prevent future mistakes where this function (`timeoutFetch()`, not `json()`) returns the wrong thing. For instance (strawman), it would prevent the trivial error of `return Promise.resolve({})`.
</details>
I'm not sure I agree. We're not specifying the full set of generics, we're leaving them implicit. So in this case, you'd just add a generic par

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 5 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKjuLKw0491YkQyDPZo:-686638008) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
> We'll have the change the health request anyway (to a POST) if it ever does need a payload, so I think it will just mislead someone looking at the codebase.
The difference is, those use the request to manipulate the URL. Changing the health request would have no effect, but might make you think you were having an effect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 5 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKju_PXQkckbPzJl-Gm:58997886) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> I'm not sure I agree. We're not specifying the full set of generics, we're leaving them implicit. So in this case, you'd just add a generic par
</details>
Comment was cut off accidentally.

Anyway I added it in, but I really don't think it confers any benefit. It doesn't give me an error if I leave off the Promise<VersionList> bit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjW2VCFf2RJxx8IR6w:-KKk86buOY7FO7TfM9VI:1833307380) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/containers/banner/outdatedBanner.tsx#L70)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Maybe, but let's track this in a separate issue. Right now the purpose is to port over the outdatedBanner. I don't think listing out a bunch of version tags or nodes will scale well, especially if the number is large.
> 
> https://github.com/cockroachdb/cockroach/issues/7343
</details>
Separate issue sounds good.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKk8CSXZhCGSUAN6-V-:-KKk8CSXZhCGSUAN6-V0:2042178600) ([raw file](https://github.com/cockroachdb/cockroach/blob/afe9a3b094a91e9c2d8f8634e5d5781a8893be1c/ui/next/app/containers/banner/outdatedBanner.tsx#L83)):*
> ```XML
>       // Check if the banner was dismissed in the last day.
>       (this.props.versionCheckDismissedFetched && !this.props.versionCheckDismissed || (Date.now() - this.props.versionCheckDismissed > MILLISECONDS_IN_DAY)) &&
>       // Check if it would show any text (ie only if there's a version mismatch or a newer version.)
> ```

nit: i.e.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 1 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKk8NzfFVnRK4sxOltz:-KKk8NzfFVnRK4sxOlu-:-1341821032) ([raw file](https://github.com/cockroachdb/cockroach/blob/afe9a3b094a91e9c2d8f8634e5d5781a8893be1c/ui/next/app/interfaces/cockroachlabs.d.ts#L1)):*
> ```XML
> // based on the types in https://github.com/cockroachlabs/registration/blob/master/db.go
> ```

Can you make this a proper comment like other files have in their headers?

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKk8VXdVuM5NuqCruCO:-1449183662) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
> any changes on the backend to the Health proto would still require changes to api.ts.

Agreed; this is fine.

Does `getDatabaseList` want the same treatment?

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKk8pK-lr2zkTNL5OUg:1277108063) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Comment was cut off accidentally.
> 
> Anyway I added it in, but I really don't think it confers any benefit. It doesn't give me an error if I leave off the Promise<VersionList> bit.
</details>
Which promise bit? I think it doesn't give you an error because this function's return type is inferred (which I think it shouldn't be, we should specify `Promise<R>`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 6 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[ui/next/app/containers/banner/outdatedBanner.tsx, line 83 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKk8CSXZhCGSUAN6-V-:-KKneGBuu9Nx_pgzEmUG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/afe9a3b094a91e9c2d8f8634e5d5781a8893be1c/ui/next/app/containers/banner/outdatedBanner.tsx#L83)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: i.e.
</details>
Done.

---

*[ui/next/app/interfaces/cockroachlabs.d.ts, line 1 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKk8NzfFVnRK4sxOltz:-KKnfYb_Ie1de3BmcYwl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/afe9a3b094a91e9c2d8f8634e5d5781a8893be1c/ui/next/app/interfaces/cockroachlabs.d.ts#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Can you make this a proper comment like other files have in their headers?
</details>
Done.

---

*[ui/next/app/util/api.ts, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTkJXUU7t23oxBpkQ:-KKnfdxZsihHg4nhPlU_:1488682622) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed585ed1a908ebd70044c3ab5ba11b65f55c5a9f/ui/next/app/util/api.ts#L161)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> > any changes on the backend to the Health proto would still require changes to api.ts.
> 
> Agreed; this is fine.
> 
> Does `getDatabaseList` want the same treatment?
</details>
Good call. Done.

---

*[ui/next/app/util/cockroachlabsAPI.ts, line 13 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-KKjTwuSbaPVKIE6f7QB:-KKnfumUtyhzif1o50Nv:-596636222) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8617a76d82df90f7646525fe0a7394f56f71e02/ui/next/app/util/cockroachlabsAPI.ts#L13)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Which promise bit? I think it doesn't give you an error because this function's return type is inferred (which I think it shouldn't be, we should specify `Promise<R>`).
</details>
Sorry. I meant, if I leave off the `Promise<VersionList>` from the versionCheck function, the type implicitly gets set to `Promise<{}>`. We may need to make the linter slightly stricter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-:-KKniDc40RK6IIk6Puii:1417088721)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
LGTM

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7339#-:-KKnjYlOnqE3Ck2saOuV:676834233)*
<!-- Sent from Reviewable.io -->
